File: workspace/src/my_py_pkg/my_py_pkg/minimal_server.py
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts

class MinimalSubscriber(Node):
    """Server example that adds two integers and responds with result"""

    def __init__(self):
        """Constructor"""

        # Call the Node class constructor with the node name
        super().__init__('minimal_server')

        # Create a service (server) object
        self._srv = self.create_service(
            AddTwoInts, 
            'add_ints', 
            self._server_callback
        )

    def _server_callback(self, req, resp):
        """Responds with sum of request integers"""
        resp.sum = req.a + req.b
        self.get_logger().info(f"Received request: a={req.a}, b={req.b}")

        return resp

def main(args=None):
    """Main entrypoint"""

    # Initialize and run node
    try:
        rclpy.init()
        node = MinimalSubscriber()
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
      pass
    finally:
        if node is not None:
          node.destroy_node()
        if rclpy.ok():
          rclpy.shutdown()

if __name__ == '__main__':
    main()

File: workspace/src/my_cpp_pkg/src/minimal_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

/**
 * Server example that adds two integers
 */
class MinimalServer : public rclcpp::Node
{
 
public:

  /**
   * Constructor
   */
  MinimalServer() : Node("minimal_server")
  {
    // Create a service (server) object
    server_ = this->create_service<example_interfaces::srv::AddTwoInts>(
      "add_ints", 
      std::bind(
        &MinimalServer::server_callback, 
        this, 
        std::placeholders::_1,
        std::placeholders::_2));
  }

private:

  /**
   * Responds with sum of request integers
   */
  void server_callback(
    const example_interfaces::srv::AddTwoInts::Request::SharedPtr req,
    const example_interfaces::srv::AddTwoInts::Response::SharedPtr resp)
  {
    // Contents of response is automatically sent back to client
    resp->sum = req->a + req->b;

    // Log request
    RCLCPP_INFO(this->get_logger(), "Received request: a=%d, b=%d", 
        (int)req->a, (int)req->b);
  }

  // Declare member variables
  rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr server_;
};

/**
 * Main entrypoint
 */
int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<MinimalServer>();
  rclcpp::spin(node);
  rclcpp::shutdown();

  return 0;
}
